### PR TITLE
🎉 Add Multitask Prompt Tuning

### PR DESCRIPTION
**Pull request by mayank31398**
Adds code for the paper: https://arxiv.org/abs/2303.02861

### CODE REVIEW

**Comment by mayank31398** [comment]
@pacman100 
Can you take a look at this PR?
This adds the functionality to use Multitask Prompt Tuning as proposed in [this paper](https://arxiv.org/abs/2303.02861).

**Comment by mayank31398** [comment]
Also, a working notebook has been added @pacman100 🤗

File: src/peft/tuners/multitask_prompt_tuning.py
Comment (gkumbhat) [inline src/peft/tuners/multitask_prompt_tuning.py:116]: Hi @mayank31398 , it seems `task_ids` at this point in code are turning out to be `int` but they need to be `Tensors`  for `torch.index_select` to work ?

**Review by pacman100 — APPROVED** [review]
Thank you so much @mayank31398 for adding MultiTask Prompt Tuning 🔥! Could you please resolve the conflicts and run `make style` and `make quality` to resolve the quality issues. Post that we can go ahead and merge this PR 🤗.

**Comment by mayank31398** [comment]
Done @pacman100

**Review by pacman100 — APPROVED** [review]
Thank you @mayank31398 for iterating!

**Comment by pacman100** [comment]
Hello, after a discussion with @younesbelkada, we were thinking if it is possible for you to add a test file for this to test minimal core components only (forward, save_pretrained, generate)?

**Comment by mayank31398** [comment]
Do you want me to provide some other script?

**Comment by pacman100** [comment]
A test file similar to https://github.com/huggingface/peft/blob/main/tests/test_adaption_prompt.py

**Review by younesbelkada — COMMENTED** [review]
Hi @mayank31398 
Thanks a lot for the great addition, I second what @pacman100 said, let's try to create a new file `test_multitask_prompt_tuning.py` to make sure nothing will get broken by future PRs, similarly as : https://github.com/huggingface/peft/blob/main/tests/test_adaption_prompt.py 
Please let us know if you need any help designing the tests!

**Comment by mayank31398** [comment]
Hey, sorry for the late reply.
Sure, Ill work on this :)

**Comment by pacman100** [comment]
Hello Mayank, please let us know if you need any help from us. Looking forward to merging this PR soon

**Comment by mayank31398** [comment]
Hey, sorry had been side-tracked with some urgent work.
Will look into doing this sometime this week :)

**Comment by mayank31398** [comment]
@pacman100 @younesbelkada really sorry for the late reply here.
I have added unit tests and everything seems to work fine.

**Comment by mayank31398** [comment]
Should we merge this?
This is a long-standing PR :)

File: src/peft/peft_model.py
Comment (younesbelkada) [inline src/peft/peft_model.py:1266]:
```suggestion
                    prompts = self.get_prompt(batch_size=batch_size, task_ids=kwargs.pop("task_ids", None))
```

Comment (younesbelkada) [inline src/peft/peft_model.py:1266]:
```suggestion
                    prompts = self.get_prompt(batch_size=batch_size, task_ids=kwargs.pop("task_ids", None))
```

**Comment by mayank31398** [comment]
Done @younesbelkada :)

**Review by BenjaminBossan — COMMENTED** [review]
Thanks so much for this PR and all the work you put in.

I haven't read the paper in detail, so I cannot comment on whether the implementation is faithful or not.

I have a few minor comments, please check my comments. Not all of that is critical. My main concern is that, if I'm not mistaken, only one `prompt_tuning_init` argument is tested. Therefore, I would like to see the tests extended to cover the other types.

As a nice-to-have feature, it would great if there were a few explanations in the notebook about what's going on + a link to the paper.

**Comment by BenjaminBossan** [comment]
I just noticed that there isn't any addition to the docs for MPT. Could you please add some documentation there? Or are we fine with adding it in a separate PR?

**Review by pacman100 — COMMENTED** [review]
Hello @mayank31398, Thank you for iterating! I believe the PR can be merged once the conflicts are resolved as there is an example showing usage and related tests. Thank you for all the work and for your kind patience 🤗

**Comment by mayank31398** [comment]
Hey @pacman100 let me add some comments in the code.
There is a some stuff that isn't clear.
I will also resolve the merge conflicts over the weekend

File: src/peft/tuners/multitask_prompt_tuning.py
Comment (mayank31398) [reply src/peft/tuners/multitask_prompt_tuning.py:124]: I doubt its needed.
+ these are not really implemented in the paper :)

File: src/peft/tuners/multitask_prompt_tuning.py
Comment (mayank31398) [reply src/peft/tuners/multitask_prompt_tuning.py:49]: rank is a parameter just like LoRA
I think it might depend a lot on the complexity of the tasks.
Hard to come up with any values.

**Comment by mayank31398** [comment]
@pacman100 I have addressed the suggestions.
Can you run the unittests?
Thanks

**Comment by mayank31398** [comment]
@pacman100 @younesbelkada lets merge this?

**Comment by mayank31398** [comment]
Hey, guys still waiting for an update :)

File: examples/image_classification/image_classification_timm_peft_lora.ipynb
Comment (younesbelkada) [inline examples/image_classification/image_classification_timm_peft_lora.ipynb:1]: Can you undo the changes related to this file? 🙏

File: examples/multilayer_perceptron/multilayer_perceptron_lora.ipynb
Comment (younesbelkada) [inline examples/multilayer_perceptron/multilayer_perceptron_lora.ipynb:1]: Same comment as above 🙏

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for your patience @mayank31398 and sorry for the delay , I left two tiny comments, apart from that I trust other maintainers review. Can you please revert the changes in the unrelated files?
Thanks!

File: examples/image_classification/image_classification_timm_peft_lora.ipynb
Comment (mayank31398) [reply examples/image_classification/image_classification_timm_peft_lora.ipynb:1]: Reverted
I think this was done by black :)

File: examples/multilayer_perceptron/multilayer_perceptron_lora.ipynb
Comment (mayank31398) [reply examples/multilayer_perceptron/multilayer_perceptron_lora.ipynb:1]: done

**Comment by mayank31398** [comment]
@younesbelkada addressed the comments

**Comment by mayank31398** [comment]
> @mayank31398 Just to let you know: We'll probably soon merge #807 which adds a new folder structure to the tuners. If you want to, you can already adjust your code to use the new structure. Otherwise, it's also fine as is, then we'll do the refactoring later.

I prefer leaving as is for now.
I am a little busy with some other stuff :)
Lets refactor later

**Review by pacman100 — APPROVED** [review]
Thank you @mayank31398 for iterating multiple times and for your patience, LGTM! 🔥🚀✨

**Comment by junzhang-zj** [comment]
@mayank31398 Why can't I locate the distillation process of the source prompt in the code? Am I missing something? 😭

**Comment by mayank31398** [comment]
@junzhang-zj yeah, it doesn't have that.
Its not simple to add it to this repo I think.
This is a basic MPT pipeline where you pretrain a joing prompt on a bunch of task, use it for initialization of target prompt and then finetune it for the target prompt.

**Comment by junzhang-zj** [comment]
@mayank31398 Thank you for your answer, which addressed my concern about carelessness!

**Pull request merged**
